### ISSUE TITLE
resource/aws_launch_template: Getting "unexpected EOF" and "connection is shut down" when applying

### ISSUE THREAD

**Issue by yzargari** [terraform-providers/terraform-provider-aws#6757]
Hello,
I'm trying to replace aws_launch_configuration with aws_launch_template on all my autoscaling groups. This means that I have to add ~150 launch templates.
The execution is failing on me every time I attempt it, unless I target just few resources (smaller plan).

The error's I'd usually get is unexpected EOF on the resource it failed to create and then connection is shut down on all the remaining resources. 

I'm running Terraform 0.11.10 with the AWS provider version 1.50.0. The only theory I have is related to the length of user_data, which in aws_launch_template takes it based64encoded. Is it possible that the length of this user_data breaks the API request?

Thanks,
Yosi

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bflad** [comment]
Hi @yzargari 👋 Sorry for the trouble. `unexpected EOF` errors indicate a code issue in the Terraform AWS provider code. We'll need some more information before we can do any troubleshooting.

Terraform should be creating a `crash.log` file in this scenario. Can you share that please (e.g. via a [Gist](https://gist.github.com))? If not, try searching for the stack trace starting with `panic: ` in any other Terraform logs if possible. See also: https://www.terraform.io/docs/internals/debugging.html

**Comment by yzargari** [comment]
It didn't create a crash.log but I did find the panic message: https://gist.github.com/yzargari/6fbbcea434405aa5ddfc51dc4f670e04

**Comment by yzargari** [comment]
Is there any workaround I can try? This is really blocking me right now and I'm willing to try any ugly hack...

**Comment by bflad** [comment]
@mskucherawy can you please open a separate bug report for that panic? Its in a different resource and will require separate triage. Thanks.

**Comment by mskucherawy** [comment]
Wilco.  (You're right, sorry, noticed this after I posted it.)

**Comment by bflad** [comment]
@yzargari 👋 Hi again. Sorry for the delayed response. Did you eventually figure out this issue? From the code lines in the panic report, it appears to be related to an "empty" configuration block for `iam_instance_profile`, e.g.

```hcl
resource "aws_launch_template" "example" {
  # ... other configuration ...
  iam_instance_profile {}
}
```

That configuration block should be omitted if you do not wish to configure IAM instance profiles with the EC2 Launch Template or should contain [either an `arn` or `name` configuration inside of it](https://www.terraform.io/docs/providers/aws/r/launch_template.html#instance-profile), e.g.

```hcl
resource "aws_launch_template" "example" {
  # ... other configuration ...
  iam_instance_profile {
    name = "${aws_iam_instance_profile.example.name}"
  }
}
```

I have submitted a pull request for the resource to skip these empty configurations instead of crashing (Terraform will report a perpetual difference for the attribute until the issue is resolved with the configuration): #7134
